gcovr/writer/html.py

```diff
@@ -367,34 +384,47 @@ def print_html_report(covdata: CovData, output_file, options):
         data["css_link"] = css_link
 
     data["theme"] = options.html_theme
+    data["row_sort_key"], data["row_sort_reverse"] = row_sort_keys(options)
 
     root_info.set_coverage(covdata)
 
     # Generate the coverage output (on a per-package basis)
     # source_dirs = set()
     files = []
-    dirs = []
     filtered_fname = ""
-    keys = sort_coverage(
+    sorted_keys = sort_coverage(
         covdata,
         show_branch=False,
         filename_uses_relative_pathname=True,
         by_num_uncovered=options.sort_uncovered,
         by_percent_uncovered=options.sort_percent,
     )
+
+    if options.html_cascaded_directories:
+        root_info.subdirs = DirectoryCoverage.from_covdata(
+            covdata, sorted_keys, options.root_filter
+        )
+        DirectoryCoverage.collapse_subdirectories(
+            root_info.subdirs, options.root_filter
+        )
+
     cdata_fname = {}
     cdata_sourcefile = {}
-    for f in keys:
+    for f in list(covdata.keys()) + list(root_info.subdirs.keys()):
```

Comment:
This list isn't sorted anymore.

Reply:
That's right, it's not sorted any more. We cannot rely on sorting at this level when managing files deeper in the directory hierarchy.

For example, not all data for all subdirectories are aggregated at this point. For example, you can't sort data for files like "A/B/fileC.cpp," when "A/B/C/fileD.cpp" hasn't been set up yet for the entry for "A/B/C"

Sorting is now handled at the Jinja level, since we are prepared with this information for Jinja rendering. Jinja has all the data needed to sort correctly. I would suggest using Jinja based sorting for the other formats also, for the future.

Reply:
> Sorting is now handled at the Jinja level, since we are prepared with this information for Jinja rendering. Jinja has all the data needed to sort correctly. I would suggest using Jinja based sorting for the other formats also, for the future.

As I wrote in https://github.com/gcovr/gcovr/pull/687#discussion_r1008750922, I think the sorting shall be changed in a separate PR. Also the sorting is changed here since you implemented a multilevel sort and changed the order if `--sort-percentage` is given.
https://github.com/gcovr/gcovr/blob/cdce979872c61b89363e997e8b6bf1f5dcc132aa/gcovr/writer/html.py#L612-L623

In general it's better to sort the keys even if it's not relevant for the result. If there is a problem on several files it's always better to abort on the same key than on a random key.

Reply:
(... I don't think you meant to use the word "abort" here, did you? )

Reply:
Abort as a result of an exception or coding error.